squirrel/plugin.py

```diff
@@ -18,30 +18,65 @@ def load_module():
         return importlib.import_module(f'squirrel.plugins.{project_type}')
 
     @staticmethod
-    def get_files(path):
-        find_output = subprocess.run(
-            f'find {path} -type f -not -path "*/[@.]*"',
-            shell=True,
-            capture_output=True,
-            text=True)
-        return find_output.stdout.strip().split('\n')
+    def get_files(path, ignores):
+        """Function to find all non-ignored files i project directory"""
+        ignores_ext = tuple(i[1:] for i in ignores.get('ext'))
+        ignores_file = tuple(ignores.get('file'))
+        ignores_dir = tuple(i[:-1] for i in ignores.get('dir_full'))
+        project_files = []
+        for file in glob('**/*', recursive=True):
+            file_path = os.path.join(path, file)
+            if not os.path.isdir(file_path):
+                if not file_path.endswith(ignores_ext) \
+                        and file_path.rsplit('/', 1)[1] not in ignores_file \
+                        and file_path.rsplit('/', 1)[0] not in ignores_dir:
+                    project_files.append(file_path)
```

Comment:
In terms of performance, we will probably never beat using the `find` command as a process. That said if we really have to use a pythonic way to traverse the tree structure and `walk` is slow as you have said, we should probably use `iglob` in that case because it's an iterator.

Reply:
I did a `time` test with `find` against `walk` when I first wrote the function, `walk` was the fastest when they had to exclude files. When it just had to list all files `find` was the fastest.
I'll change `glob` with `iglob`. But I need to help/tip on how to remove all `if not in` statments if it's posible

Reply:
Update:
```python
ignore = ['*.tmp', 'test.txt', '<root>/tmp/*']

# Have to define the variable inside [] to set/tuple or else it doesn´t work,
# even if I define it to set/tuple before I pass it to iglob
for file in iglob('**/*[!f"{set(ignores)}"]'):
```
This one works, but it has some bugs I havn't sorted out yet.
- filenames with multiple `.` doesn´t get ignored e.g `something.test.txt`
- files in `./tmp` is ignored but folder is added. 
  - Can be fixed with an `.isdir()` expresion
  - Maybe not a problem to pass a folder to `get_count()`?

Tried to pass `<root>/tmp/*` to ignore_pattern in `Handler()` but it doesn't work. Maybe keep full directory as it is and use a `if not in` also in `iglob`?

Reply:
As I see it there two conditions that a file must have to be add to the list:
1. it's not in an ignored directory
2. it's not an ignored file

So one solution might be to first filter the `iglob` with a `from_ignored_dir` fucntion and then with a `ignored_file`.
```Python
filter = lambda file: True if not from_ignored_dir(file) and not ignored_file(file) else False
files = [ file for file in iglob('**/*', recursive=True) if filter(file)]
```
As globing patters are very limited, I don't think they work in our case here.

Reply:
How about this one?

Hence sets don't allow duplicates all ignored files is removed from `project_files`.
Havn't tried it runing `watch` yet, but in pythonshell it works.

```python
    def get_files(path, ignores):
        """Function to find all non-ignored files i project directory"""
        ignore = ignores.get('ignore_path')
        project_files = set()
        for file in iglob('**/*', recursive=True):
            if os.path.isfile(file):
                file_path = os.path.join(path, file)
                project_files.add(file_path)
        project_files.difference_update(ignore)
        return project_files
```

Reply:
Looks good to me. However, I would add a test to verify that a file is not in an ignored directory.

Reply:
The files in ignored directorys are already in the ignored set. As import_ignores was called just before this function.
After this the Handler() function takes over checking if a modified file is in the ignore patterns while `watch` is running.

That is if the suggestions in import_ignores with recursion doesn't change that?
Have to look at it tomorrow. GN

Reply:
Yes, all the files inside the dir will be ignored but in that case if new files are added to the directory they won't be ignored.

Reply:
```python
    def not_ignored_folder(self, file):
        if f'{"".join(file.rsplit("/", 1)[:-1])}/' \
                not in self.ignores.get('dir'):
            return True
```

In that case this function inside `Handler()` takes care of it, thats why we still need `ignores['dir']`.

Reply:
Yeah! I'm sorry you're right. okay I'll test things a bit more.

Reply:
No problem. It works as I belive it should on my machine.